test/cpp/end2end/xds/xds_override_host_end2end_test.cc

```diff
@@ -0,0 +1,241 @@
+// Copyright 2023 gRPC authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <chrono>
+#include <string>
+#include <thread>
+#include <vector>
+
+#include <gmock/gmock.h>
+#include <gtest/gtest.h>
+
+#include "absl/strings/str_format.h"
+#include "absl/strings/str_join.h"
+#include "absl/strings/str_split.h"
+
+#include "src/core/ext/filters/client_channel/backup_poller.h"
+#include "src/core/lib/gprpp/match.h"
+#include "src/proto/grpc/testing/xds/v3/cluster.grpc.pb.h"
+#include "src/proto/grpc/testing/xds/v3/outlier_detection.grpc.pb.h"
+#include "src/proto/grpc/testing/xds/v3/router.grpc.pb.h"
+#include "src/proto/grpc/testing/xds/v3/stateful_session.grpc.pb.h"
+#include "src/proto/grpc/testing/xds/v3/stateful_session_cookie.pb.h"
+#include "test/core/util/scoped_env_var.h"
+#include "test/cpp/end2end/xds/xds_end2end_test_lib.h"
+
+namespace grpc {
+namespace testing {
+namespace {
+using ::envoy::extensions::filters::http::stateful_session::v3::StatefulSession;
+using ::envoy::extensions::filters::network::http_connection_manager::v3::
+    HttpFilter;
+using ::envoy::extensions::http::stateful_session::cookie::v3 ::
+    CookieBasedSessionState;
+
+class OverrideHostTest : public XdsEnd2endTest {
+ protected:
+  struct Cookie {
+    std::string value;
+    std::set<std::string> attributes;
+    std::string raw;
+  };
+
+  static std::map<std::string, Cookie, std::less<>> ParseCookies(
```

Comment:
The way this is written, it requires passing in all of the metadata, and it returns and parses every single cookie, which seems like overkill when the caller just wants one cookie.  Instead, I suggest something like this:

```
static Cookie ParseCookie(
    absl::string_view cookie_name,
    absl::string_view set_cookie_header_value);
```

The caller can find the `set-cookie` header and pass the value of that header in, and the function can return just the one specific cookie that the caller wants.